src/v/cloud_storage/topic_manifest_downloader.h

```diff
@@ -0,0 +1,64 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#pragma once
+
+#include "base/outcome.h"
+#include "cloud_storage/remote.h"
+#include "cloud_storage/remote_label.h"
+#include "cloud_storage/topic_manifest.h"
+#include "model/fundamental.h"
+
+namespace cloud_storage {
+
+enum class find_topic_manifest_outcome {
+    success = 0,
+
+    // There are no topic manifests belonging to a given topic.
+    no_matching_manifest,
+
+    // Indicates that multiple topic manifests are found for the given topic
+    // and a hint must be provided to determine winner.
+    multiple_matching_manifests,
+};
+
+// Encapsulates downloading manifests for a given topic.
+//
+// Topic manifests have gone through a few format/naming schemes:
+// - cluster-uuid-labeled, binary format
+// - hash-prefixed, binary format
+// - hash-prefixed, JSON format
+//
+// This downloader returns the topic manifest, hiding the details of this
+// history from callers.
+class topic_manifest_downloader {
+public:
+    topic_manifest_downloader(
+      const cloud_storage_clients::bucket_name bucket,
+      std::optional<ss::sstring> hint,
+      const model::topic_namespace topoic,
+      remote& remote);
+
+    // Attempts to download the topic manifest, transparently checking paths
+    // that have been used by older versions of Redpanda if a new manifest
+    // isn't found.
+    ss::future<result<find_topic_manifest_outcome, error_outcome>>
```

Comment:
Comment here could reiterate the order of attempted downloads, as illustrated in function comments and commit message.

Reply:
Added this in the class header. I think it's unimportant for callers of this method to understand.